marlowe-object/test/Language/Marlowe/Object/TypesSpec.hs

```diff
@@ -0,0 +1,46 @@
+{-# LANGUAGE GADTs #-}
+
+module Language.Marlowe.Object.TypesSpec (spec) where
+
+import Data.Aeson (Result (..), ToJSON (..), fromJSON)
+import Data.Binary (Binary, decode, encode)
+import Data.Foldable (traverse_)
+import Data.Proxy (Proxy (..))
+import Language.Marlowe.Object.Gen ()
+import Language.Marlowe.Object.Types
+import Test.Hspec
+import Test.Hspec.QuickCheck (prop)
+import Test.QuickCheck hiding (Success, label)
+import Test.QuickCheck.Classes (Laws (..), eqLaws, jsonLaws, ordLaws, showLaws, showReadLaws)
+
+spec :: Spec
+spec = do
+  describe "ObjectBundle" do
+    -- Testing hand-written Binary instance only. Other instances are derived
+    -- and expensive to run.
+    checkLaws $ binaryLaws $ Proxy @ObjectBundle
+
+  describe "LabelledObject" do
+    checkLaws $ eqLaws $ Proxy @LabelledObject
+    checkLaws $ ordLaws $ Proxy @LabelledObject
+    checkLaws $ jsonLaws $ Proxy @LabelledObject
+    checkLaws $ showLaws $ Proxy @LabelledObject
+    checkLaws $ showReadLaws $ Proxy @LabelledObject
+
+  describe "Contract" do
+    prop "Can decode JSON generated from core contracts" \contract ->
+      fromJSON (toJSON contract) === Success (fromCoreContract contract)
+
+binaryLaws :: forall a. (Binary a, Show a, Arbitrary a, Eq a) => Proxy a -> Laws
+binaryLaws _ =
+  Laws
+    "Binary"
+    [
+      ( "inverse"
+      , property \(a :: a) -> decode (encode a) === a
+      )
+    ]
+
+checkLaws :: Laws -> Spec
```

Comment:
I'm concerned about the `IsString` instances in `Language.Marlowe.Object.Types` because it appears that sometimes we have the law `fromString x = read x` but other times we might have `fromString x = read ("\\"" ++ x ++ "\\"")`. This has implications for parsing and for `OverloadedStrings`. Is there or should there be a single law for `IsString` in the codebase?

Reply:
`fromString x = read x` looks wrong. I'll double-check these instances. I think that the following should hold in general for any "string-like" types:

```hs
read . show = id -- This should always be true (I know it currently is not)
Show a => Read a -- Always define read and show, and show produces a valid Haskell expression.
fromString s = read $ (showChar '"' . showString s . showChar '"') ""
```

Reply:
Here's what I've done:

```hs
instance Read Base16 where
  readPrec = do
    String s <- lexP
    Right parsed <- pure $ decodeBase16' $ T.pack s
    pure $ Base16 parsed

instance IsString Base16 where
  fromString = either (error . showString "fromString(Base16): " . T.unpack) Base16 . decodeBase16' . T.pack
```